### PR TITLE
[AMDGPU] RA inserted scalar instructions can be at the BB top

### PR DESCRIPTION
**Pull request by cdevadas**
We adjust the insertion point at the BB top for spills/copies during RA to ensure they are placed after the exec restore instructions required for the divergent control flow execution. This is, however, required only for the vector operations. The insertions for scalar registers can still go at the BB top.

### CODE REVIEW

**Review by rampitec — COMMENTED** [review]
You also need a specific small test for the situation.

File: llvm/include/llvm/CodeGen/MachineBasicBlock.h [hunk 1]
```diff
@@ -846,8 +846,10 @@ class MachineBasicBlock
 
   /// Return the first instruction in MBB after I that is not a PHI, label or
   /// debug.  This is the correct point to insert copies at the beginning of a
-  /// basic block.
-  iterator SkipPHIsLabelsAndDebug(iterator I, bool SkipPseudoOp = true);
+  /// basic block. The optional arg \p Reg would allow targets to apply some
```
[the remark this answers is not in the document]
Comment (qcolombet) [reply llvm/include/llvm/CodeGen/MachineBasicBlock.h:849]: +1 on the description not being descriptive enough.

**Comment by cdevadas** [comment]
Lit test precheckin + rebase + comments addressed.

File: llvm/include/llvm/CodeGen/MachineBasicBlock.h [hunk 2]
```diff
@@ -846,8 +846,11 @@ class MachineBasicBlock
 
   /// Return the first instruction in MBB after I that is not a PHI, label or
   /// debug.  This is the correct point to insert copies at the beginning of a
-  /// basic block.
-  iterator SkipPHIsLabelsAndDebug(iterator I, bool SkipPseudoOp = true);
+  /// basic block. \p Reg is an optional argument passed during register
+  /// allocator to have additional target specific checks for its spill/copy
```
Comment (rampitec) [inline llvm/include/llvm/CodeGen/MachineBasicBlock.h:850]: allocator -> allocation. I also believe it shall be spill/restore/copy.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 3]
```diff
@@ -8474,16 +8474,24 @@ unsigned SIInstrInfo::getLiveRangeSplitOpcode(Register SrcReg,
...
+bool SIInstrInfo::isBasicBlockPrologue(const MachineInstr &MI,
+                                       Register Reg) const {
   // We need to handle instructions which may be inserted during register
   // allocation to handle the prolog. The initial prolog instruction may have
   // been separated from the start of the block by spills and copies inserted
-  // needed by the prolog.
+  // needed by the prolog. However, the insertions for scalar registers can
+  // always be placed at the BB top as they are independent of the exec mask
+  // value.
   uint16_t Opc = MI.getOpcode();
```
[the remark this answers is not in the document]
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:8485]: Still the whole block can be avoided if Reg is unset.

File: llvm/include/llvm/CodeGen/MachineBasicBlock.h (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/MachineBasicBlock.h:850]: This is still a roundabout way of saying what the register value is. Say something that expresses this is the register being defined for a spill/split

Comment (arsenm) [reply llvm/include/llvm/CodeGen/MachineBasicBlock.h:850]: I'm thinking this should be a differently named function, specifically for the spill insert point

**Comment by cdevadas** [comment]
Review comments addressed.

File: llvm/include/llvm/CodeGen/MachineBasicBlock.h [hunk 4]
```diff
@@ -846,8 +846,10 @@ class MachineBasicBlock
 
   /// Return the first instruction in MBB after I that is not a PHI, label or
   /// debug.  This is the correct point to insert copies at the beginning of a
-  /// basic block.
-  iterator SkipPHIsLabelsAndDebug(iterator I, bool SkipPseudoOp = true);
+  /// basic block. \p Reg is the register being defined for a spill/split during
```
Comment (rampitec) [inline llvm/include/llvm/CodeGen/MachineBasicBlock.h:849]: Actually spill does not define a register, it kills it. Restore defines register.

**Review by rampitec — COMMENTED** [review]
LGTM with a small nit.

**Review by rampitec — APPROVED** [review]
LGTM

**Pull request merged**